src/NSwag.Generation.AspNetCore/AspNetCoreOpenApiDocumentGenerator.cs

```diff
@@ -127,15 +127,51 @@ public async Task<OpenApiDocument> GenerateAsync(ApiDescriptionGroupCollection a
 
         /// <summary>Gets the default serializer settings representing System.Text.Json.</summary>
         /// <returns>The settings.</returns>
-        public static JsonSerializerSettings GetSystemTextJsonSettings()
+        public static JsonSerializerSettings GetSystemTextJsonSettings(IServiceProvider serviceProvider)
         {
             // If the ASP.NET Core website does not use Newtonsoft.JSON we need to provide a 
             // contract resolver which reflects best the System.Text.Json behavior.
             // See https://github.com/RicoSuter/NSwag/issues/2243
-            return new JsonSerializerSettings
+            JsonSerializerSettings settings = new JsonSerializerSettings
             {
                 ContractResolver = new CamelCasePropertyNamesContractResolver()
             };
+
+#if NETSTANDARD2_0
+            if (serviceProvider == null)
+            {
+                return settings;
+            }
+                
+            dynamic options = null;
+            try
+            {
+                var optionsAssembly = Assembly.Load(new AssemblyName("Microsoft.AspNetCore.Mvc.Core"));
+                var optionsType = typeof(IOptions<>).MakeGenericType(optionsAssembly.GetType("Microsoft.AspNetCore.Mvc.JsonOptions", true));
+                options = serviceProvider?.GetService(optionsType) as dynamic;
+            }
+            catch
+            {
+            }
+
+            if (!(options?.Value?.JsonSerializerOptions is System.Text.Json.JsonSerializerOptions jsonOptions))
```

Comment:
Isn't the JsonOptions always registered, even when Newtonsoft stuff is enabled?

Reply:
JsonOptions is always registered from AspNetCore 3.0, it was MvcJsonOptions before.
Since this is currently called as a fallback when Newtonsoft options are not found, JsonOptions should always be registered when it gets to this point. Assuming Newtonsoft options are always found when running on AspNetCore pre 3.0